zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1098 @@
+
+        // An empty checkpoint list can't actually verify any blocks.
+        match checkpoints.keys().cloned().next() {
+            None => {
+                return Err("there must be at least one checkpoint, for the genesis block".into())
+            }
+            Some(BlockHeight(0)) => {}
+            _ => return Err("checkpoints must start at the genesis block height 0".into()),
+        };
+
+        Ok(CheckpointVerifier {
+            checkpoint_list: checkpoints,
+            queued: <BTreeMap<BlockHeight, QueuedBlockList>>::new(),
+            // We start by verifying the genesis block, by itself
+            current_checkpoint_range: Some((Included(BlockHeight(0)), Included(BlockHeight(0)))),
+        })
+    }
+
+    /// Return the block height of the highest checkpoint in the checkpoint list.
+    ///
+    /// If there is only a single checkpoint, then the maximum height will be
+    /// zero. (The genesis block.)
+    ///
+    /// The maximum height is constant for each checkpoint list.
+    ///
+    /// Returns None if there are no checkpoints.
+    fn get_max_checkpoint_height(&self) -> Option<BlockHeight> {
+        self.checkpoint_list.keys().cloned().next_back()
+    }
+
+    /// Return the most recently verified checkpoint height.
+    ///
+    /// The height increases as blocks are verified.
+    ///
+    /// If verification has not started yet, returns None.
+    /// If verification has finished, returns the maximum checkpoint height.
+    fn get_previous_checkpoint_height(&self) -> Option<BlockHeight> {
+        match self.current_checkpoint_range {
+            Some((Included(BlockHeight(0)), _)) => None,
+            Some((Excluded(height), _)) => Some(height),
+            None => self.get_max_checkpoint_height(),
+            _ => unreachable!(),
+        }
+    }
+
+    /// Return the next checkpoint height that we want to verify.
+    ///
+    /// The height increases as blocks are verified.
+    ///
+    /// If verification has not started yet, returns zero (the genesis block).
+    /// If verification has finished, returns None.
+    fn get_next_checkpoint_height(&self) -> Option<BlockHeight> {
+        match self.current_checkpoint_range {
+            Some((_, Included(height))) => Some(height),
+            None => None,
+            _ => unreachable!(),
+        }
+    }
+
+    /// Return the most recently verified checkpoint's hash.
+    ///
+    /// If verification has not started yet, returns None. To verify the genesis
+    /// block, check that the parent block hash is all zeroes. (Bitcoin "null".)
+    ///
+    /// If verification has finished, returns the maximum checkpoint's hash.
+    fn get_previous_checkpoint_hash(&self) -> Option<BlockHeaderHash> {
+        match self.get_previous_checkpoint_height() {
+            // Every checkpoint height must have a hash
+            Some(height) => Some(*self.checkpoint_list.get(&height).unwrap()),
```

Comment:
```suggestion
            Some(height) => Some(*self.checkpoint_list.get(&height).expect("Every checkpoint height must have a hash")),
```

Reply:
Fixed all instances of this pattern in adfa7adc8ab634c59252ee01e3fcacf7a8012a38